File: proxy.py
#!/usr/bin/env python

import tornado.web
import tornado.ioloop
import tornado.httpclient
# from urlparse import urlparse

<<<<<<< HEAD
=======

>>>>>>> bb39f124284664757142637387fd839422e2accb
disallowed_response_headers = frozenset([
    'connection',
    'keep-alive',
    'proxy-authenticate',
    'proxy-authorization',
    'te',
    'trailers',
    'transfer-encoding',
    'upgrade',
    # all above are hop-by-hop headers
    'content-encoding',  # may cause decoding errors
    'content-length',
])


class ProxyHandler(tornado.web.RequestHandler):
    SUPPORTED_METHODS = ['GET', 'POST']

<<<<<<< HEAD
    def handle_response(self, response):
=======
    def _handle_response(self, response):
>>>>>>> bb39f124284664757142637387fd839422e2accb
        self.set_status(response.code)

        # check the last comment of http://goo.gl/4w5yj
        for h in response.headers.keys():
            if h.lower() not in disallowed_response_headers:
                list_values = response.headers.get_list(h)
                if len(list_values) == 1:
                    self.set_header(h, list_values[0])
                else:
                    for v in list_values:
                        self.add_header(h, v)

        if response.body:
            self.write(response.body)

        self.finish()

    @tornado.web.asynchronous
    def get(self):
        http_req = tornado.httpclient.HTTPRequest(
            url=self.request.uri,
            method=self.request.method,
            body=self.request.body,
            headers=self.request.headers,
            follow_redirects=False,
            allow_nonstandard_methods=True
        )

<<<<<<< HEAD
        http_client = tornado.httpclient.AsyncHTTPClient()

        try:
            http_client.fetch(http_req, self.handle_response)
        except tornado.httpclient.HTTPError as err:
            self.handle_response(err.response)
=======
        http_client_1 = tornado.httpclient.AsyncHTTPClient()
        http_client_2 = tornado.httpclient.AsyncHTTPClient()

        try:
            http_client_1.fetch(http_req, self._handle_response)
            http_client_2.close()
        except tornado.httpclient.HTTPError as err:
            self._handle_response(err.response)
        except Exception, err:
            self.set_status(500)
            self.write('Internal Server Error: ' + str(err))
            self.finish()

        try:
            http_client_2.fetch(http_req, self._handle_response)
            http_client_1.close()
        except tornado.httpclient.HTTPError as err:
            self._handle_response(err.response)
>>>>>>> bb39f124284664757142637387fd839422e2accb
        except Exception, err:
            self.set_status(500)
            self.write('Internal Server Error: ' + str(err))
            self.finish()

<<<<<<< HEAD
    # define self.post() as self.get()
=======
    # def self.post() as self.get()
>>>>>>> bb39f124284664757142637387fd839422e2accb
    post = get


application = tornado.web.Application(
    [(r'.*', ProxyHandler)],
    debug=True  # please set to false in production environments
)

if __name__ == "__main__":
    application.listen(8888)
    tornado.ioloop.IOLoop.instance().start()

File: proxy.js
var http = require('http');
var url = require('url');
 
var proxy_server = http.createServer(function(client_request, client_response) {
    var parsed_url = url.parse(client_request.url);
    var proxy_request_options = {
        host: parsed_url.host,
        path: parsed_url.path,
        method: client_request.method,
        headers: client_request.headers
    };

    var flag = 0;
    var proxy_request_1, proxy_request_2;
 
    proxy_request_1 = http.request(proxy_request_options, function(proxy_response_1) {
        proxy_response_1.on('data', function(chunk) {
            if (flag === 1) {
                client_response.write(chunk);
            }
        });
        proxy_response_1.on('end', function() {
            if (flag === 1) {
                client_response.end();
            }
        });

        flag = 1;
        // proxy_request_2.abort();
        console.info('Agent 1 ' + client_request.method + ' ' + client_request.url);
        client_response.writeHead(proxy_response_1.statusCode, proxy_response_1.headers);
    });
    proxy_request_1.on('error', function(err) {
        // aborting connections will cause many ECONNRESET errors
        // if (err.code !== 'ECONNRESET') {
            console.error('Proxy Request 1 Error: ' + err.message);
        // }
    });

    proxy_request_2 = http.request(proxy_request_options, function(proxy_response_2) {
        proxy_response_2.on('data', function(chunk) {
            if (flag === 2) {
                client_response.write(chunk);
            }
        });
        proxy_response_2.on('end', function() {
            if (flag === 2) {
                client_response.end();
            }
        });

        flag = 2;
        // proxy_request_1.abort();
        console.info('Agent 2 ' + client_request.method + ' ' + client_request.url);
        client_response.writeHead(proxy_response_2.statusCode, proxy_response_2.headers);
    });
    proxy_request_2.on('error', function(err) {
        // if (err.code !== 'ECONNRESET') {
            console.error('Proxy Request 2 Error: ' + err.message);
        // }
    });


    client_request.pipe(proxy_request_1);
    client_request.pipe(proxy_request_2);
    client_request.on('error', function(err) {
        console.error('Server Error: ' + err.message);
    });
}).listen(8080, function() {
    console.info('Listening on 8080');
});

process.on('uncaughtException', function(err) {
    console.error('Caught exception: ' + err);
});
